MerkleMountainRange/src/merklemountainrange.rs

```diff
@@ -0,0 +1,322 @@
+    pub fn get_peak(&self) -> u32 {
+        let mut height_counter = 0;
+        let mmr_len = self.get_last_added_index() as i128;
+        while (mmr_len - i128::pow(2, height_counter + 1) - 2) >= 0 {
+            height_counter += 1;
+        }
+        height_counter
+    }
+
+    /// This function adds a vec of leaf nodes to the mmr.
+    pub fn add_vec(&mut self, objects: Vec<T>) {
+        for object in objects {
+            self.add_single(object);
+        }
+    }
+
+    /// This function adds a new leaf node to the mmr.
+    pub fn add_single(&mut self, object: T) {
+        let node_hash = object.get_hash();
+        let node = MerkleNode::new(node_hash);
+        self.mmr.push(node);
+        if is_node_right(self.get_last_added_index()) {
+            self.data.insert(node_hash, object);
+            self.add_single_no_leaf(self.get_last_added_index())
+        }
+    }
+
+    // This function adds non leaf nodes, eg nodes that are not directly a hash of data
+    // This is iterative and will continue to up and till it hits the top, will be a future left child
+    fn add_single_no_leaf(&mut self, index: u64) {
+        let new_hash =
+            self.data.get(&self.mmr[index as usize].hash).unwrap().concat(self.mmr[peer_index(index) as usize].hash);
+
+        let new_node = MerkleNode::new(new_hash);
+        self.mmr.push(new_node);
+        if is_node_right(self.get_last_added_index()) {
+            self.add_single_no_leaf(self.get_last_added_index())
+        }
+    }
+
+    // This function is just a private function to return the index of the last added node
+    fn get_last_added_index(&self) -> u64 {
+        (self.mmr.len() - 1) as u64
+    }
+}
+/// This function takes in the index and calculates the index of the peer.
+pub fn peer_index(index: u64) -> u64 {
+    let index_count = u64::pow(2, index as u32 + 1) - 1;
+    if is_node_right(index) {
+        index - index_count
+    } else {
+        index + index_count
+    }
+}
+
+/// This function takes in the index and calculates if the node is the right child node or not.
+/// If the node is the tree root it will still give the answer as if it is a child of a node.
+/// This function is an iterative function as we might have to subtract the largest left_most tree.
+pub fn is_node_right(index: u64) -> bool {
+    let mut height_counter = 0;
+    while (index as i128 - i128::pow(2, height_counter + 1) - 2) >= 0 {
+        height_counter += 1;
+    }
+    if (index - u64::pow(2, height_counter + 1) - 2) == 0 {
+        return false;
+    };
+    let cloned_index = index - u64::pow(2, height_counter + 1) - 1;
+    if (cloned_index - u64::pow(2, height_counter + 1) - 2) == 0 {
+        return true;
+    };
+    is_node_right(cloned_index)
+}
+
+/// This function takes in the index and calculates the height of the node
+/// This function is an iterative function as we might have to subtract the largest left_most tree.
+pub fn get_node_height(index: u64) -> u32 {
```

Comment:
This function also needs a test. It also seems very similar to the one above, maybe we can reuse some of the code

Reply:
It only reuses the math to " remove the largest part of the tree", but I believe that trying to reuse this single math function is not worth the extra lines its going to add.